### ISSUE TITLE
enc_password generate function

### ISSUE THREAD

**Issue by AleksandrTan** [itsMoji/Instagram_SSL_Pinning#96]
Good day. 
Please tell me if someone knows the encryption algorithm for the parameter "enc_password" when requested 
https://i.instagram.com/api/v1/accounts/login/?(Private API)
I would like to implement it on Python.
I tried using the function from here https://github.com/dilame/instagram-private-api/issues/1010 but it doesn't work.
I get an answer {"message": "There was an error with your request. Please try again.", "status": "fail"}.
Thanks you.

**Comment by kingbotss** [comment]
Generate function same function in python would help.
`function encryptPassword(password: string): { time: string, encrypted: string } {
  const plainKey = Buffer.from(ig.state.passwordEncryptionPubKey, 'base64').toString();
  const randKey = crypto.randomBytes(32);
  const iv = crypto.randomBytes(12);
  const resEncrypted = crypto.publicEncrypt(plainKey, randKey);
  const cipher = crypto.createCipheriv('aes-256-gcm', randKey, iv);
  const time = Math.floor(Date.now() / 1000).toString();
  cipher.setAAD(Buffer.from(time));
  const aesEncrypted = concat(cipher.update(password, 'utf8'), cipher.final());
  const sizeBuffer = Buffer.alloc(2, 0);
  sizeBuffer.writeInt16LE(resEncrypted.byteLength, 0);
  const authTag = cipher.getAuthTag();
  return {
    time,
    encrypted: concat(
      Buffer.from([1, ig.state.passwordEncryptionKeyId]),
      iv,
      sizeBuffer,
      resEncrypted, aesEncrypted, authTag)
      .toString('base64'),
  };
}`

You need crypto library with aes-256-gcm support.

**Comment by AleksandrTan** [comment]
Thank you very much)
I'll try this option.

**Comment by AleksandrTan** [comment]
Unfortunately this code hasn't worked yet.
I ran it "pure" that is, as Node code, but it doesn't work.
I also tried the function code from this repository(https://github.com/dilame/instagram-private-api/blob/master/src/repositories/account.repository.ts), but it did not give a positive result either.

**Comment by AleksandrTan** [comment]
Perhaps I am doing something wrong with the requests, but on the other hand, if I deliberately pass the wrong parameter - "username", I will get the following answer - {'message': "We can't find an account with "wrong_username". Try another phone number or email, or if you don't have an Instagram account, you can sign up.", 'invalid_credentials': True, 'error_title': "Can't Find Account", 'buttons': [{'title': 'Sign Up', 'action': 'switch_to_signup_flow'}, {'title': 'Try Again', 'action': 'dismiss'}], 'status': 'fail', 'error_type': 'invalid_user'}

**Comment by Nerixyz** [comment]
> I ran it "pure" that is, as Node code, but it doesn't work.

That's funny because when I do the exact same it just works.

**Comment by AleksandrTan** [comment]
> > I ran it "pure" that is, as Node code, but it doesn't work.
> 
> That's funny because when I do the exact same it just works.

How long ago did you run this code?

**Comment by AleksandrTan** [comment]
APP_VERSION = '177.0.0.30.119'
APP_VERSION_CODE = '276028050'
I am using these options.

**Comment by AleksandrTan** [comment]
const crypto = require('crypto')
function encryptPassword(password, key_pub, key_id){
    const plainKey = Buffer.from(key_pub, 'base64').toString();
    const randKey = crypto.randomBytes(32);
    const iv = crypto.randomBytes(12);
    const resEncrypted = crypto.publicEncrypt(plainKey, randKey);
    const cipher = crypto.createCipheriv('aes-256-gcm', randKey, iv);
    const time = Math.floor(Date.now() / 1000).toString();
    cipher.setAAD(Buffer.from(time));
    const aesEncrypted = Buffer.concat([cipher.update(password, 'utf8'), cipher.final()]);
    const sizeBuffer = Buffer.alloc(2, 0);
    sizeBuffer.writeInt16LE(resEncrypted.byteLength, 0);
    const authTag = cipher.getAuthTag();
    return "#PWD_INSTAGRAM:4:"+time+":"+Buffer.concat([Buffer.from([1, key_id]), iv, sizeBuffer, resEncrypted, aesEncrypted, authTag]) .toString('base64')
}

**Comment by kingbotss** [comment]
> APP_VERSION = '177.0.0.30.119'
> APP_VERSION_CODE = '276028050'
> I am using these options.

I am using this version. It is working fine.

**Comment by AleksandrTan** [comment]
Could you briefly describe your login request algorithm?
My path is like this:
1. I make 4 pre-requests to the api:
- qe/sync/ (get key string 'ig-set-password-encryption-pub-key' and key id 'ig-set-password-encryption-key-id' )
- launcher/sync/
- accounts/get_prefill_candidates/
- accounts/contact_point_prefill/
2. Then I take the cookie from the last one, generate the headers, the parameter string and send the request:
  - accounts/login/
  **Headings:**
{'Accept-Encoding': 'gzip, deflate',
 'Accept-Language': 'en_US',
 'Connection': 'close',
 'Content-Length': '1133',
 'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
 'Cookie': 'mid=YFM21wABAAGChGwhnriosny-rOVt; '
           'csrftoken=DnkwaylZVmCc2GxSo4vcSvqKc30kbMcL',
 'Host': 'i.instagram.com',
 'IG-INTENDED-USER-ID': '0',
 'User-Agent': 'Instagram 177.0.0.30.119 Android (26/8.0.0; 420dpi; 1080x2094; '
               'samsung; SM-N950F; greatlte; samsungexynos8895; en_US; '
               '276028050)',
 'X-Bloks-Is-Layout-RTL': 'false',
 'X-Bloks-Version-Id': '38e807d1f50024907c1026934e57bf28a7c421e5ffcdc0a7b0aa31dbd44acd74',
 'X-FB-HTTP-Engine': 'Liger',
 'X-IG-Android-ID': 'android-49e885c1bd66fc43',
 'X-IG-App-ID': '567067343352427',
 'X-IG-App-Locale': 'en_US',
 'X-IG-Bandwidth-Speed-KBPS': '6987.417',
 'X-IG-Bandwidth-TotalBytes-B': '0',
 'X-IG-Bandwidth-TotalTime-MS': '0',
 'X-IG-Capabilities': '3brTvwM=',
 'X-IG-Connection-Type': 'WIFI',
 'X-IG-Device-ID': '89c119ee-d056-46f2-b391-13db8e5200a1',
 'X-IG-Device-Locale': 'en_US',
 'X-IG-Mapped-Locale': 'en_US',
 'X-IG-WWW-Claim': '0',
 'X-MID': 'YFM21wABAAGChGwhnriosny-rOVt',
 'X-Pigeon-Rawclienttime': '1616484109.658',
 'X-Pigeon-Session-Id': '32642387-21ee-453d-a6b2-576c76848446',
 'X-Tigon-Is-Retry': 'True'}
**Params:**
{'_csrftoken': 'DnkwaylZVmCc2GxSo4vcSvqKc30kbMcL',
 'adid': '',
 'country_codes': '[{"country_code":"1","source":["default"]}]',
 'device_id': 'android-55da46e7b387bce5',
 'enc_password': '#PWD_INSTAGRAM:4:1616484129:ARFTKrJg7B7M2yiUMJwAAQBL/4UD1ox8Zf6Trqip2j9iQ1MRcqJiSrsBzbOWHEJmTCdZ8F9LJ+AGUj+mKIkwLfxGZJ+PfwZrn1/nxcz2KJp1+fBS/WbJ8y+El/0swsEXrrtF+hJsx9VtJrLAKyztVKv3QqZGqyEEwiIhyWmlUXV58mjSwQq2t3TZLmFQUCeaCTQOikwLP9QZ0wn44DrAEil3F/VAPAo0/KNAUydDcH/TVoIDlbJbi2O2PoPZ1ot6bThEqdv0LVFLY/IfF9IGGYHDfGW9zHadfk9maC5coEkaQ6aNNZ5guD63U/sRJXy7F6Bxnx6ZiXzm6inqJJKhilHziil2FLld8DCYe7olpxGelI5cyyqwVTZ7ij76llVn7nI+dBM5Xszh',
 'google_tokens': '[]',
 'guid': '89c119ee-d056-46f2-b391-13db8e5200a1',
 'jazoest': '22430',
 'login_attempt_count': 1,
 'phone_id': 'e36dcef1-a72b-489e-9942-57bd70b164d1',
 'username': 'username'}
I generate a string from the parameters:
signed_body=SIGNATURE.%7B%22jazoest%22%3A%20%2222676%22%2C%20%22country_codes%22%3A%20%22%5B%7B%5C%22country_code%5C%22%3A%5C%221%5C%22%2C%5C%22source%5C%22%3A%5B%5C%22default%5C%22%5D%7D%5D%22%2C%20%22phone_id%22%3A%20%22e9bc27ec-8dbd-4172-89db-d86a5daf77bb%22%2C%20%22enc_password.......
Thanks you.

**Comment by kingbotss** [comment]
Same function from Instagram-private-api
